### PR TITLE
EBL-230 - add is_active flag to apollo environments

### PR DESCRIPTION
**Pull request by matvey-mtn**

### CODE REVIEW

File: apollo-backend/src/test/java/io/logz/apollo/EnvironmentTest.java [hunk 1]
```diff
@@ -67,4 +67,97 @@ public void testGetAllEnvironments() throws ApolloClientException {
...
+
+        // Add environments
+        Environment testEnvironment = createAndSubmitEnvironment(apolloTestClient, false);
+        Environment activeTestEnvironment = createAndSubmitEnvironment(apolloTestClient);
+
+        // Get all environments should return everything
+        List<Environment> allEnvironments = apolloTestClient.getAllEnvironments();
+
+        boolean foundNotActiveEnvironment = allEnvironments.stream()
+                .anyMatch(environment -> environment.getId().equals(testEnvironment.getId()));
```
Comment (mayhuberman) [inline apollo-backend/src/test/java/io/logz/apollo/EnvironmentTest.java:84]: Maybe put the assert right after the check

File: apollo-backend/src/test/java/io/logz/apollo/EnvironmentTest.java [hunk 2]
```diff
@@ -67,4 +67,97 @@ public void testGetAllEnvironments() throws ApolloClientException {
...
+        foundNotActiveEnvironment = allActiveEnvironments.stream()
+                .anyMatch(environment -> environment.getId().equals(testEnvironment.getId()));
+        assertThat(foundNotActiveEnvironment).isFalse();
+
+        // Active environment should be present in response
+        Optional<Environment> environmentFromApi = allActiveEnvironments.stream()
+                .filter(environment -> environment.getId().equals(activeTestEnvironment.getId()))
+                .findFirst();
+
+        assertThat(environmentFromApi).isPresent();
```
Comment (mayhuberman) [inline apollo-backend/src/test/java/io/logz/apollo/EnvironmentTest.java:104]: Why here checking the presence of the environment and not like the checks before?

File: apollo-backend/src/test/java/io/logz/apollo/EnvironmentTest.java [hunk 3]
```diff
@@ -67,4 +67,97 @@ public void testGetAllEnvironments() throws ApolloClientException {
...
+
+        // Make sure we got the correct one
+        assertThat(environmentFromApi.get().getName()).isEqualTo(activeTestEnvironment.getName());
```
Comment (mayhuberman) [inline apollo-backend/src/test/java/io/logz/apollo/EnvironmentTest.java:107]: If it's the same id then it's the same environment, no need to all of these checks :)

File: apollo-backend/src/test/java/io/logz/apollo/EnvironmentTest.java [hunk 4]
```diff
@@ -67,4 +67,97 @@ public void testGetAllEnvironments() throws ApolloClientException {
...
+
+        ApolloTestClient apolloTestClient = Common.signupAndLogin();
+
+        // Add environment
+        Environment testEnvironment = createAndSubmitEnvironment(apolloTestClient);
+
+        // Check that environment is created
+        boolean foundTestEnvironment = apolloTestClient.getAllEnvironments().stream()
+                .anyMatch(environment -> environment.getId().equals(testEnvironment.getId()));
+        assertThat(foundTestEnvironment).isTrue();
```
Comment (mayhuberman) [inline apollo-backend/src/test/java/io/logz/apollo/EnvironmentTest.java:127]: There is a different test for that, "testAddAndGetEnvironment", no need for this check here :)

File: apollo-backend/src/test/java/io/logz/apollo/EnvironmentTest.java [hunk 5]
```diff
@@ -67,4 +67,97 @@ public void testGetAllEnvironments() throws ApolloClientException {
...
+
+        // Updated environment should be present in response
+        Optional<Environment> environmentFromApi = apolloTestClient.getAllEnvironments().stream()
+                .filter(environment -> environment.getId().equals(newEnvironment.getId()))
+                .findFirst();
+
+        assertThat(environmentFromApi).isPresent();
+
+        // Make sure we got the correct one
+        assertThat(environmentFromApi.get().getName()).isEqualTo(newEnvironment.getName());
```
Comment (mayhuberman) [inline apollo-backend/src/test/java/io/logz/apollo/EnvironmentTest.java:154]: The same check of all the environment's members exist in "testGetAllEnvironments".
Maybe export the logic from there to a new method, and use there and here :)

File: apollo-client/src/main/java/io/logz/apollo/clients/ApolloClient.java [hunk 6]
```diff
@@ -79,6 +101,11 @@ public Environment getEnvironment(int id) throws ApolloClientException {
         });
     }
 
+    public List<Environment> getAllActiveEnvironments() throws ApolloClientException {
+        return genericApolloClient.getResult("/environment?active=true", new TypeReference<List<Environment>>() {
```
Comment (mayhuberman) [inline apollo-client/src/main/java/io/logz/apollo/clients/ApolloClient.java:105]: Same as before, maybe not clear enough and better with two different endpoints

File: ui/app/scripts/services/apolloApiService.js [hunk 7]
```diff
@@ -28,6 +28,12 @@ function ApiService($q, $http){
         return $http.get(CONFIG.appUrl + 'environment/');
     };
 
+    var getAllActiveEnvironments = function() {
+        return $http.get(CONFIG.appUrl + 'environment/', {
+            params: {active: true}
```
Comment (mayhuberman) [inline ui/app/scripts/services/apolloApiService.js:33]: Same as before 😬

File: apollo-client/src/main/java/io/logz/apollo/clients/GenericApolloClient.java [hunk 8]
```diff
@@ -188,8 +188,9 @@ String getUserName() {
 
     private String getFullUrlWithToken(String url) {
         String urlPrefix = prefix.isPresent() ? "/" + prefix.get() : "";
-        String tokenPostfix = StringUtils.isNotBlank(token) ? "?_token=" + token : "";
-        return protocol + "://" + hostname + ":" + port + urlPrefix + url + tokenPostfix;
+        String tokenPrefix = url.contains("?") ? "&" : "?";
```
Comment (mayhuberman) [inline apollo-client/src/main/java/io/logz/apollo/clients/GenericApolloClient.java:191]: Why? What is the use?

Comment (matvey-mtn) [reply apollo-client/src/main/java/io/logz/apollo/clients/GenericApolloClient.java:191]: if someone wants to send query params
for example `/api?x=1&y=2`
without this fix resulting request will look like the following: `/api?x=1&y=2?_token=****`
making it invalid. question mark appears twice in this request.

File: apollo-backend/src/test/java/io/logz/apollo/DeploymentTest.java [hunk 9]
```diff
@@ -209,4 +214,16 @@ public void testDeploymentWithAdditionalParams() throws Exception {
         Fabric8TestMethods.assertIngressLabelExists(apolloToKubernetes.getKubernetesIngress(),
                 RealDeploymentGenerator.DEFAULT_INGRESS_LABLE_KEY, ingressValue);
     }
+
+    @Test
+    public void testDeploymentToInactiveEnvironment() throws Exception {
+        exceptionRule.expect(Exception.class);
+        exceptionRule.expectMessage("Cannot deploy. Target environment is not active.");
```
Comment (mayhuberman) [inline apollo-backend/src/test/java/io/logz/apollo/DeploymentTest.java:221]: Important test!
You can use assertThatThrownBy() instead of this

**Pull request merged**